coordinator/persistence-cassandra-4.1/src/main/java/io/stargate/db/cassandra/impl/StargateSystemKeyspace.java

```diff
@@ -0,0 +1,126 @@
+import org.apache.cassandra.cql3.statements.SelectStatement;
+import org.apache.cassandra.db.SystemKeyspace;
+import org.apache.cassandra.db.virtual.*;
+import org.apache.cassandra.dht.IPartitioner;
+import org.apache.cassandra.dht.Token.TokenFactory;
+import org.apache.cassandra.locator.InetAddressAndPort;
+import org.apache.cassandra.schema.*;
+import org.apache.cassandra.transport.ProtocolVersion;
+import org.apache.cassandra.utils.FBUtilities;
+import org.apache.cassandra.utils.MurmurHash;
+
+public class StargateSystemKeyspace extends VirtualKeyspace {
+  public static final String SYSTEM_KEYSPACE_NAME = "stargate_system";
+  public static final String LOCAL_TABLE_NAME = "local";
+  public static final String LEGACY_PEERS_TABLE_NAME = "peers";
+  public static final String PEERS_V2_TABLE_NAME = "peers_v2";
+
+  public static final UUID SCHEMA_VERSION = UUID.fromString("17846767-28a1-4acd-a967-f609ff1375f1");
+
+  public static final StargateSystemKeyspace instance = new StargateSystemKeyspace();
+
+  private StargateSystemKeyspace() {
+    super(
+        SYSTEM_KEYSPACE_NAME,
+        ImmutableList.of(
+            new StargatePeersView(), new StargateLegacyPeersView(), new StargateLocalView()));
+  }
+
+  private final ConcurrentMap<InetAddressAndPort, StargatePeerInfo> peers =
+      new ConcurrentHashMap<>();
+  private final StargateLocalInfo local = new StargateLocalInfo();
+
+  public ConcurrentMap<InetAddressAndPort, StargatePeerInfo> getPeers() {
+    return peers;
+  }
+
+  public StargateLocalInfo getLocal() {
+    return local;
+  }
+
+  public void persistLocalMetadata() {
+    local.setClusterName(DatabaseDescriptor.getClusterName());
+    local.setReleaseVersion(FBUtilities.getReleaseVersionString());
+    local.setCqlVersion(QueryProcessor.CQL_VERSION);
+    local.setNativeProtocolVersion(String.valueOf(ProtocolVersion.CURRENT.asInt()));
+    local.setDataCenter(DatabaseDescriptor.getLocalDataCenter());
+    local.setRack(DatabaseDescriptor.getEndpointSnitch().getLocalRack());
+    local.setPartitioner(DatabaseDescriptor.getPartitioner().getClass().getName());
+    local.setBroadcastAddress(FBUtilities.getJustBroadcastAddress());
+    local.setListenAddress(FBUtilities.getJustLocalAddress());
+    local.setNativeAddress(FBUtilities.getJustBroadcastNativeAddress());
+    local.setNativePort(DatabaseDescriptor.getNativeTransportPort());
+    local.setHostId(SystemKeyspace.getOrInitializeLocalHostId());
+    local.setTokens(
+        StargateSystemKeyspace.generateRandomTokens(
+            FBUtilities.getBroadcastNativeAddressAndPort(), DatabaseDescriptor.getNumTokens()));
+  }
+
+  public static boolean isSystemPeers(CQLStatement statement) {
+    if (statement instanceof SelectStatement) {
+      SelectStatement selectStatement = (SelectStatement) statement;
+      return selectStatement.keyspace().equals(SchemaConstants.SYSTEM_KEYSPACE_NAME)
+          && selectStatement.table().equals(SystemKeyspace.LEGACY_PEERS);
```

Comment:
C*4 persistence only seems to check keyspace, not table name for some reason. But it is probably better to actually verify table name than to rely there not being anything.